### PR TITLE
[clang][FatLTO] Avoid UnifiedLTO until it can support WPD/CFI

### PR DESCRIPTION
**Pull request by ilovepi**
Currently, the UnifiedLTO pipeline seems to have trouble with several
LTO features, like SplitLTO units, which means we cannot use important
optimizations like Whole Program Devirtualization or security hardening
instrumentation like CFI.

This patch reverts FatLTO to using distinct pipelines for Full LTO and
ThinLTO. It still avoids module cloning, since that was error prone.

### CODE REVIEW

File: llvm/lib/Passes/PassBuilderPipelines.cpp [hunk 1]
```diff
@@ -1531,22 +1531,31 @@ PassBuilder::buildPerModuleDefaultPipeline(OptimizationLevel Level,
...
+  if (PGOOpt && PGOOpt->Action == PGOOptions::SampleUse) {
+    if (ThinLTO) {
+      // Use the ThinLTO post-link pipeline with sample profiling
+      MPM.addPass(
+          buildThinLTODefaultPipeline(Level, /*ImportSummary=*/nullptr));
+      return MPM;
+    }
+    MPM.addPass(SampleProfileLoaderPass(PGOOpt->ProfileFile,
+                                        PGOOpt->ProfileRemappingFile,
+                                        ThinOrFullLTOPhase::FullLTOPostLink));
```
Comment (nikic) [inline llvm/lib/Passes/PassBuilderPipelines.cpp:1552]: Don't think we need this one -- SampleProfileLoad during pre-link is only skipped in ThinLTO per https://github.com/llvm/llvm-project/blob/2e0cd6077a756fa28023d907045e6e53c50ec732/llvm/lib/Passes/PassBuilderPipelines.cpp#L1001-L1006 So for non-thin LTO it should have happened already.

**Review by nikic — APPROVED** [review]
LGTM, this seems like the pragmatic choice if we want a working FatLTO implementation in LLVM 18.

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 2]
```diff
@@ -1046,6 +1047,7 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
         MPM.addPass(PrintModulePass(*OS, "", CodeGenOpts.EmitLLVMUseLists,
                                     /*EmitLTOSummary=*/true));
       }
+
```
[the remark this answers is not in the document]
Comment (ilovepi) [reply clang/lib/CodeGen/BackendUtil.cpp:1050]: Thank you!

File: llvm/lib/Passes/PassBuilderPipelines.cpp (hunk 1 above)
Comment (ilovepi) [reply llvm/lib/Passes/PassBuilderPipelines.cpp:1552]: oh, you're right. This can basically just go back to the way it was then. Thanks for catching that.

**Comment by nikic** [comment]
There's a merge conflict (probably after https://github.com/llvm/llvm-project/pull/79041).

**Comment by ilovepi** [comment]
Since this has passed Linux premerge checks and I think Windows CI is having some issues I tested this locally on a Windows machine and its passing. I'm going to go ahead and merge.

**Pull request merged**